src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
+                        }
+
+                        OP_PICK
+                        | OP_ROLL => {
+                            // (xn ... x2 x1 x0 n - xn ... x2 x1 x0 xn)
+                            // (xn ... x2 x1 x0 n - ... x2 x1 x0 xn)
+                            if stack.size() < 2 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            let n =
+                                u16::try_from(ScriptNum::new(stack.top(-1)?, require_minimal, None)?)
+                                .map_err(|_| ScriptError::InvalidStackOperation)?;
+                            stack.pop()?;
+                            if usize::from(n) >= stack.size() {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            let vch: ValType =
+                                stack.top(-isize::try_from(n).map_err(|_| ScriptError::InvalidStackOperation)? - 1)?
+                                .clone();
+                            if op == OP_ROLL {
+                                stack.erase(stack.end() - usize::from(n) - 1, None);
+                            }
+                            stack.push_back(vch)
+                        }
+
+                        OP_ROT => {
+                            // (x1 x2 x3 -- x2 x3 x1)
+                            //  x2 x1 x3  after first swap
+                            //  x2 x3 x1  after second swap
+                            if stack.size() < 3 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            stack.swap(-3, -2)?;
+                            stack.swap(-2, -1)?;
+                        }
+
+                        OP_SWAP => {
+                            // (x1 x2 -- x2 x1)
+                            if stack.size() < 2 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            stack.swap(-2, -1)?;
+                        }
+
+                        OP_TUCK => {
+                            // (x1 x2 -- x2 x1 x2)
+                            if stack.size() < 2 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            let vch = stack.top(-1)?.clone();
+                            stack.insert(stack.end() - 2, vch)
+                        }
+
+
+                        OP_SIZE => {
+                            // (in -- in size)
+                            if stack.size() < 1 {
+                                return set_error(ScriptError::InvalidStackOperation);
+                            }
+                            let bn =
+                                ScriptNum::try_from(stack.top(-1)?.len()).map_err(|_| ScriptError::PushSize)?;
```

Comment:
Is the error behaviour the same as C++ here?

Reply:
I believe it’s actually unreachable, as `ScriptError::PushSize` is what happens when a value larger than 520 bytes (`MAX_SCRIPT_ELEMENT_SIZE`) is pushed onto the stack in the first place. So `stack.top(_)?.len()` should always fit into the `i64` that `ScriptNum` holds, or `ScriptError::PushSize` would have happened already.

Should I change this to `unreachable!` with a similar explanation?

Reply:
You can use `.expect` in such cases, e.g.
```suggestion
                                ScriptNum::try_from(stack.top(-1)?.len()).expect('stack element size fits in ScriptNum");
```

Reply:
A :confused: emoji indicates that this is *not* addressed in #197. (It's not a great choice of emoji, there just isn't a better one in the supported set.)